Detectors/TOF/calibration/src/TOFCalibCollector.cxx

```diff
@@ -0,0 +1,147 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "TOFCalibration/TOFCalibCollector.h"
+#include "Framework/Logger.h"
+#include <cassert>
+#include <iostream>
+#include <sstream>
+#include <TStopwatch.h>
+
+namespace o2
+{
+namespace tof
+{
+
+using Slot = o2::calibration::TimeSlot<o2::tof::TOFCalibInfoSlot>;
+
+//_____________________________________________
+void TOFCalibInfoSlot::fill(const gsl::span<const o2::dataformats::CalibInfoTOF> data)
+{
+  // fill container
+  // we do not apply any calibration at this stage, it will be applied when we
+  // process the data before filling the CCDB in the separate process
+
+  for (int i = data.size(); i--;) {
+    auto ch = data[i].getTOFChIndex();
+    auto offset = std::accumulate(mEntriesSlot.begin(), mEntriesSlot.begin() + ch, 0);
+    mTOFCollectedCalibInfoSlot.emplace(mTOFCollectedCalibInfoSlot.begin() + offset, data[i].getTimestamp(), data[i].getDeltaTimePi(), data[i].getTot(), data[i].getFlags());
+    mEntriesSlot[ch]++;
+  }
+}
+
+//_____________________________________________
+void TOFCalibInfoSlot::merge(const TOFCalibInfoSlot* prev)
+{
+  // merge data of 2 slots
+
+  int addedPerChannel = 0;
+  int offset = 0;
+
+  LOG(DEBUG) << "Merging two slots with entries: current slot -> " << mTOFCollectedCalibInfoSlot.size() << " , previous slot -> " << prev->mTOFCollectedCalibInfoSlot.size();
+  for (int ch = 0; ch < Geo::NCHANNELS; ch++) {
+    offset += mEntriesSlot[ch];
+    if (prev->mEntriesSlot[ch] == 0)
+      continue;
+    auto offsetprevStart = addedPerChannel;
+    auto offsetprevEnd = addedPerChannel + prev->mEntriesSlot[ch];
+    addedPerChannel += prev->mEntriesSlot[ch];
+    mTOFCollectedCalibInfoSlot.insert(mTOFCollectedCalibInfoSlot.begin() + offset, prev->mTOFCollectedCalibInfoSlot.begin() + offsetprevStart, prev->mTOFCollectedCalibInfoSlot.begin() + offsetprevEnd);
```

Comment:
``vector.insert`` is very expensive operation, moreover do it in loop for so many channels. It will be much faster to copy piecewise the mTOFCollectedCalibInfoSlot vectors of current and prev. slot to temporary vector, then to do 
mTOFCollectedCalibInfoSlot.swap(tmpVec);

BTW, isn't there a risk to have too many entries in the vector? Perhaps one should set some limit on the number of entries per channel?

Reply:
Thanks for the suggestion, I will implement it as you suggest.
As for the number of entries, in principle we dump everything either at every timeframe, or when we reach a certain number of entries (per channel or absolute), so there should be no such a risk. 
BTW, what would be the limit of a vector?

Reply:
Actually, I am not sure I understand the suggestion. The function allows to add to the current slot channel by channel the information of the previous slot, channel by channel. If `insert` is too expensive, I can build a new vector filling it for every channel with the entries of the current slot, then of the previous one, etc. E.g.:
``
for (int ch = 0; ch < Geo::NCHANNELS; ch++) {
    for ... loop over the entries in the current vector for channel ch... {
       add these entries to a new vector
    }
     for ... loop over the entries in the previous vector for channel ch... {
       add these entries to a new vector
    }
}
then yes, one has to swap the new vector with the one of the current slot.

Reply:
this actually what I've suggested: build a new vector for increasing channels. But for that you will not to have the data sorted in channelID, right?

Reply:
I am sorry, I don't understand the questions: the vectors mTOFCollectedCalibInfoSlot of the slots are ordered in channel: first all entries in channel 0, then all for channel 1, then channel 2... etc.

Reply:
yes, but also the data should be ordered in channels, in order to incrementally build joint vector.

Reply:
Again I don't understand: here I am merging two slots, which were already filled, and they are filled ordered in channel. What data are you referring to?

Reply:
Sorry, in the previous reply I've confused this piece with ``fill(..`` method I was commenting above.

Reply:
Done